### PR TITLE
Issues #94 and #89 adding image sharing and image preview 

### PR DESCRIPTION
**Pull request by tskull01**

### CODE REVIEW

File: client/src/utils/imagePicker.js [hunk 1]
```diff
@@ -3,8 +3,9 @@ const imagePicker = (e) => {
   return new Promise((resolve, reject) => {
     let reader = new FileReader();
     let file = e.target.files[0];
-
-    reader.readAsDataURL(file);
+    if (file) {
+      reader.readAsDataURL(file);
+    }
```
Comment (muke1908) [inline client/src/utils/imagePicker.js:8]:
```suggestion
    if (!file) {
        reject('No file selected')
    }
     reader.readAsDataURL(file);
```

File: client/src/components/Messaging/Message.js [hunk 2]
```diff
@@ -29,12 +29,14 @@ export const Message = ({ handleSend, index, message: { owner, body, local } })
   }, []);
 
   const [darkMode] = useContext(ThemeContext);
+  const regex = /\bdata:image\b/g;
+  const imgTest = body.match(regex);
```
Comment (muke1908) [inline client/src/components/Messaging/Message.js:33]:
```suggestion
  const isImg = body.match(regex);
```

File: client/src/components/Messaging/Message.js [hunk 3]
```diff
@@ -29,12 +29,14 @@ export const Message = ({ handleSend, index, message: { owner, body, local } })
...
   return (
     <div className={owner === true ? styles.messageRight : styles.messageLeft}>
       <div className={styles.messageInfo}>
         <div className={styles.sentReceived}>You {owner === true ? 'sent' : 'received'}</div>
         <div className={`${styles.messageContainer} ${!darkMode && styles.lightModeContainer}`}>
-          {body}
+          {imgTest ? <Image src={body} maxWidth="100%" maxHeight="auto" /> : body}
```
Comment (muke1908) [inline client/src/components/Messaging/Message.js:39]:
```suggestion
          {isImg ? <Image src={body} maxWidth='300px' maxHeight='300px' /> : body}
```

File: client/src/components/Messaging/NewMessageForm.js [hunk 4]
```diff
@@ -29,8 +38,24 @@ export const NewMessageForm = ({ handleSubmit, text, setText }) => {
...
+            className={styles.inputImagePicker}
+            type="file"
+            onChange={async (e) => {
+              let name = e.target.files[0].name;
+              let imgUrl = await imagePicker(e);
+              if (imgUrl) {
+                setPreviewImage(true);
+                setSelectedImage(imgUrl);
+                setText(name);
+              }
```
Comment (muke1908) [inline client/src/components/Messaging/NewMessageForm.js:51]: Let's create a separate function and call it here in `onChange`.

File: client/src/pages/messaging/index.js [hunk 5]
```diff
@@ -79,7 +79,7 @@ const Chat = () => {
     myKeyRef.current = _keyPair;
   };
 
-  const handleSubmit = (e) => {
+  const handleSubmit = (e, imgUrl = '') => {
```
Comment (muke1908) [inline client/src/pages/messaging/index.js:82]: It looks like design inconsistency. 
Either keep both text and imageUrl at the parent state or, pass both of them in the function argument.

File: client/src/pages/messaging/index.js [hunk 6]
```diff
@@ -91,13 +91,23 @@ const Chat = () => {
...
+      );
+    } else {
+      setMessages((prevMsg) =>
+        prevMsg.concat({
+          body: text,
+          sender: userId,
+          local: true
+        })
+      );
+    }
```
Comment (muke1908) [inline client/src/pages/messaging/index.js:110]:
```suggestion
setMessages((prevMsg) =>
        prevMsg.concat({
          body: imgUrl || text,
          sender: userId,
          local: true
        })
      );
```

**Review by muke1908 — CHANGES_REQUESTED** [review]

File: client/src/components/Messaging/NewMessageForm.js [hunk 7]
```diff
@@ -29,8 +38,24 @@ export const NewMessageForm = ({ handleSubmit, text, setText }) => {
       />
       <div className={styles.imagePickerContainer}>
```
Comment (muke1908) [inline client/src/components/Messaging/NewMessageForm.js:39]: I would suggest making a separate component for imagePicker - which will pick and show preview

**Comment by muke1908** [comment]
Looks like you have deleted `.env.sample`

**Comment by tskull01** [comment]
It looks like PubNub errors on any message over 32 KB. From their forums they recommend either sending a URL that points to the image or sending the message in multiple parts. 

https://support.pubnub.com/support/solutions/articles/14000043783-can-i-send-large-messages-with-pubnub-

**Comment by muke1908** [comment]
> It looks like PubNub errors on any message over 32 KB. From their forums they recommend either sending a URL that points to the image or sending the message in multiple parts.
> 
> https://support.pubnub.com/support/solutions/articles/14000043783-can-i-send-large-messages-with-pubnub-

Hi @tskull01, yes we eventually have to upload the image separately with API. Let's merge this for now.
The idea of sharing image is https://github.com/muke1908/chat-e2ee/wiki/Idea:-Secure-image-sharing You can take a look

**Pull request merged**